cardano-cli/src/Cardano/CLI/Shelley/Run/SingleAddressWallet.hs

```diff
@@ -0,0 +1,205 @@
+import           Cardano.Config.Types (SocketPath)
+
+import qualified Cardano.Crypto.Hash.Class   as Crypto
+
+import           Cardano.Slotting.Slot (SlotNo (..), fromWithOrigin)
+
+import           Control.Monad.Trans.Except (ExceptT)
+import           Control.Monad.Trans.Except.Extra (firstExceptT, hoistMaybe)
+
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+
+import           Ouroboros.Consensus.Block (CodecConfig)
+import           Ouroboros.Consensus.Shelley.Ledger (ShelleyBlock)
+import           Ouroboros.Consensus.Shelley.Protocol.Crypto (TPraosStandardCrypto)
+
+import           Ouroboros.Network.Block (getTipPoint, pointSlot)
+
+import qualified Shelley.Spec.Ledger.Coin as Shelley (Coin (..))
+import qualified Shelley.Spec.Ledger.LedgerState as Shelley (genesisId, minfee)
+import qualified Shelley.Spec.Ledger.PParams as Shelley (PParams)
+import qualified Shelley.Spec.Ledger.TxData as Shelley (Addr (..), TxId (..), TxIn (..), TxOut (..))
+import qualified Shelley.Spec.Ledger.UTxO as Shelley (UTxO (..))
+
+-- | Build a basic transaction that consists of a single input and output
+-- and covers its own fees.
+--
+-- For the input, we query the node's local state for a UTxO that is
+-- sufficient to cover the min fee of the transaction that will be
+-- constructed. For the output, we simply spend the change back to the sender.
+buildSingleAddressWalletTransaction
+  :: CodecConfig (ShelleyBlock TPraosStandardCrypto)
+  -> NetworkMagic
+  -> SocketPath
+  -> Maybe SlotNo
+  -> SigningKey
+  -> ExceptT CliError IO TxSigned
+buildSingleAddressWalletTransaction _ _ _ _ (SigningKeyByron _) =
+  panic "buildSingleAddressWalletTransaction: SigningKeyByron"
+buildSingleAddressWalletTransaction codecCfg nm sockPath mbTxTtl
+                                    skey@(SigningKeyShelley shelleySkey) = do
+    -- Query the local tip from chain sync protocol
+    tipPoint <- withIOManagerE $ \iocp ->
+      liftIO $ getTipPoint <$> getLocalTip iocp codecCfg network sockPath
+
+    -- Query the PParams from the node's local state
+    pparams <- firstExceptT NodeLocalStateQueryError $
+      queryPParamsFromLocalState codecCfg network sockPath tipPoint
+
+    -- Build a transaction containing one dummy input (representing an input
+    -- that can cover the fee) and one dummy output (representing the change
+    -- output).
+    --
+    -- We then use this dummy transaction to calculate the min fee which we'll
+    -- utilize when constructing the real transaction.
+    let txTtl = chooseTxTtl tipPoint
+        dummyTxForFeeCalc = buildDummyShelleyTx txTtl
+        minFee = calculateShelleyMinFee pparams dummyTxForFeeCalc
+
+    -- Query the UTxO from the node's local state
+    utxo <- firstExceptT NodeLocalStateQueryError $
+      queryFilteredUTxOFromLocalState codecCfg network sockPath (Set.singleton addr) tipPoint
+
+    -- Find an unspent transaction output that can be used to cover the fee.
```

Comment:
should this read find unspent tx input?

Reply:
Nope, this was what I meant since an input references an unspent transaction output. However, maybe the wording here isn't great and I should revisit it.